numpy/lib/shape_base.py

```diff
@@ -89,9 +90,20 @@ def apply_along_axis(func1d, axis, arr, *args, **kwargs):
     outshape = asarray(arr.shape).take(indlist)
     i.put(indlist, ind)
     res = func1d(arr[tuple(i.tolist())], *args, **kwargs)
-    #  if res is a number, then we have a smaller output array
+
+    # see gh-8363
+    _strcase = False
+    _dtypes = set()
+
+    # If res is a number, then we have a smaller output array.
     if isscalar(res):
```

Comment:
Ok, here's my proposed fix (on top of #8441 ):

```diff
diff --git a/numpy/lib/shape_base.py b/numpy/lib/shape_base.py
index 342f47d..804ece0 100644
--- a/numpy/lib/shape_base.py
+++ b/numpy/lib/shape_base.py
@@ -10,7 +10,7 @@ from numpy.core.fromnumeric import product, reshape, transpose
 from numpy.core import vstack, atleast_3d
 from numpy.lib.index_tricks import ndindex
 from numpy.matrixlib.defmatrix import matrix  # this raises all the right alarm bells
-
+from numpy.compat.py3k import bytes, unicode
 
 __all__ = [
     'column_stack', 'row_stack', 'dstack', 'array_split', 'split',
@@ -96,11 +96,14 @@ def apply_along_axis(func1d, axis, arr, *args, **kwargs):
     # invoke the function on the first item
     ind0 = next(inds)
     res = func1d(inarr_view[ind0], *args, **kwargs)
+    res_type = type(res)
     res = asanyarray(res)
 
+    is_py_str = issubclass(res_type, (bytes, unicode))
+
     # insert as many axes as necessary to create the output
     outshape = arr.shape[:axis] + res.shape + arr.shape[axis+1:]
-    outarr = zeros(outshape, res.dtype)
+    outarr = zeros(outshape, res.dtype if not is_py_str else object)
     # matrices call reshape in __array_prepare__, and this is apparently ok!
     if not isinstance(res, matrix):
         outarr = res.__array_prepare__(outarr)
@@ -119,6 +122,9 @@ def apply_along_axis(func1d, axis, arr, *args, **kwargs):
 
     outarr = res.__array_wrap__(outarr)
 
+    if is_py_str:
+        outarr = outarr.astype(res_type)
+
     return outarr
```

Special case `str` and `bytes`, _and not dtypes_, because if the user really wants fixed width strings, then they probably returned an array scalar anyway